src/thor/expansion_action.cc

```diff
@@ -0,0 +1,188 @@
+#include "thor/worker.h"
+
+#include "baldr/json.h"
+#include "baldr/rapidjson_utils.h"
+#include "midgard/constants.h"
+#include "midgard/logging.h"
+#include "midgard/polyline2.h"
+#include "midgard/util.h"
+
+using namespace rapidjson;
+using namespace valhalla::midgard;
+
+namespace {
+using namespace valhalla;
+static std::unordered_map<const Options::ExpansionProps, const char*> kPropPaths =
+    {{Options_ExpansionProps_edge_ids, "/features/0/properties/edge_ids"},
+     {Options_ExpansionProps_statuses, "/features/0/properties/statuses"},
+     {Options_ExpansionProps_distances, "/features/0/properties/distances"},
+     {Options_ExpansionProps_durations, "/features/0/properties/durations"},
+     {Options_ExpansionProps_costs, "/features/0/properties/costs"}};
+
+} // namespace
+
+namespace valhalla {
+namespace thor {
+
+std::string thor_worker_t::expansion(Api& request) {
+  // time this whole method and save that statistic
+  measure_scope_time(request, "thor_worker_t::expansion");
+
+  // get the request params
+  auto options = request.options();
+  auto exp_action = options.expansion_action();
+  bool skip_opps = options.skip_opposites();
+  auto exp_props_pbf = options.expansion_props();
+  std::unordered_set<baldr::GraphId> opp_edges;
+  std::unordered_set<Options::ExpansionProps> exp_props_set;
+
+  // default generalization to ~ zoom level 15
+  float gen_factor = options.has_generalize() ? options.generalize() : 10.f;
+
+  // default the expansion geojson so its easy to add to as we go
+  Document dom;
+  dom.SetObject();
+  // set algorithm to Dijkstra, will be overwritten by other algos
+  SetValueByPointer(dom, "/properties/algorithm", "unidirectional_dijkstra");
+  SetValueByPointer(dom, "/type", "FeatureCollection");
+  SetValueByPointer(dom, "/features/0/type", "Feature");
+  SetValueByPointer(dom, "/features/0/geometry/type", "MultiLineString");
+  SetValueByPointer(dom, "/features/0/geometry/coordinates", Value(kArrayType));
+  SetValueByPointer(dom, "/features/0/properties", Value(kArrayType));
+  for (const auto& prop : exp_props_pbf) {
+    rapidjson::Pointer(kPropPaths[static_cast<Options_ExpansionProps>(prop)])
+        .Set(dom, Value(kArrayType));
+    exp_props_set.insert(static_cast<Options_ExpansionProps>(prop));
```

Comment:
need a container which implements some find/count or smth, PBF repeated fields seem to behave more like a vector